src/client.rs

```diff
@@ -330,7 +325,7 @@ where
     B: IntoBuf,
     B::Buf: 'static,
 {
-    type Item = (Client<B>, Connection<T, B>);
+    type Item = (Connection<T, B>, SendRequest<B>);
```

Comment:
Why the change in order? Without this change, it reads as `let (client, conn) = await!(handshake(io))`, where `client` (now a `SendRequest`) is a sender, and `conn` is a thing to push into an executor. It matches somewhat the `let (tx, rx) = channel()` pattern.

With the new change to `SendRequest`, it would be `let (tx, conn) = await!(handshake(io))`.

Reply:
I mostly changed it because I've been following the pattern of `Foo::somefn -> (Self, ...)` where the `Self` type comes first.

Another option would be to move the `handshake` fn off of the `Connection` type, which I think may be more "rusty" for fns that return a tuple.

Reply:
I see the benefits of each:

- `let (conn, tx) = Connection::handshake(io)`: Less imports, follow `Foo::foo -> (Self, ...)` pattern
- `let (tx, conn) = Connection::handshake(io)`: Less imports, follow `let (tx, rx) =` pattern.
- `let (tx, conn) = client::handshake(io)`: Possibly an extra import, the `client` module, but doesn't break the previous two patterns.

Reply:
I would do either 1) or 3).

However, I also wouldn't try to match the `(tx, rx)` order much, mostly because the connection handle isn't really a receiver.

Reply:
You know, actually, the `conn` part is almost useless anyways. Is there user going to do anything to it because push it onto an executor? ... I guess by returning the future, someone can wrap it more easily than if they had to provide a custom executor... And it does emit a connection error if they occur...